### PR TITLE
BUG: Fix buffer overflows in tokenizer.c that caused python to segfault with certain

### LINKED ISSUE #9205: pandas error kills IPython kernel
Doing this simple thing can kill an IPython 2 notebook kernel (version:GH master):

```python
url = 'http://nssdc.gsfc.nasa.gov/planetary/factsheet/index.html'
pd.read_table(url)
```
Note that I know this will fail, I just don't expect it to kill a notebook kernel?

Version:
pandas: 128ce857f1cd
IPython: 13facaf0206240a7301e045666143d68305d0119

### PR DESCRIPTION
**Pull request by selasley**
closes #9205

### CODE REVIEW

**Comment by jreback** [comment]
whoosh, quite a number of changes for a 'simple' bug.....

**Comment by selasley** [comment]
there was some pruning of routines that didn't seem to be used anymore and some changes to the way pointers were reallocated and freed while investigating the problem.  the heart of the fix is checking lengths against the buffer caps as items are added to the buffers.    i thought the extra checking might slow things down slightly, but vbench claims the tokenizer branch is as fast as upstream/master.  i'll take a look at the other csv issues you pointed me to earlier

**Comment by jreback** [comment]
@selasley hahh.. wasn't being critical. I am very glad you refactored. Sometimes just needs to be done! thanks.

File: doc/source/whatsnew/v0.16.0.txt [hunk 1]
```diff
@@ -210,3 +208,6 @@ Bug Fixes
 - Fixes issue with ``index_col=False`` when ``usecols`` is also specified in ``read_csv``. (:issue:`9082`)
 - Bug where ``wide_to_long`` would modify the input stubnames list (:issue:`9204`)
 - Bug in to_sql not storing float64 values using double precision. (:issue:`9009`)
+
+
+- Bug in csv.  Buffer overflows with certain malformed input files (:issue:`9205`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.16.0.txt]: Bug in ``read_csv``.

**Comment by jreback** [comment]
will have to test this on windows.

**Comment by selasley** [comment]
:)  I was a bit nervous about changing so many lines of other people's code.  Thanks for checking on windows.  I don't have access to a windows machine at home or work to do any testing.
I fixed the typo in the whatsnew document and pushed the changes

**Comment by jreback** [comment]
ok, this looks good on windows.

can you do a vbench run on this? just to confirm nothing has changed? 

ping when done and we'll get this in.

**Comment by jreback** [comment]
merged via https://github.com/pydata/pandas/commit/c6c9c0bfc10e24bdd0e4d32f7a8faf7fbc2294b2

**Comment by jreback** [comment]
thanks!

**Pull request closed without merging**